drivers/SmartThings/matter-thermostat/src/init.lua

```diff
@@ -240,9 +270,105 @@ local subscribed_attributes = {
 }
 
+local function epoch_to_iso8601(time)
+  return os.date("!%Y-%m-%dT%H:%M:%SZ", time)
+end
+
+local function tbl_contains(array, value)
+  for _, element in ipairs(array) do
+    if element == value then
+      return true
+    end
+  end
+  return false
+end
+
+local get_total_cumulative_energy_imported = function(device)
+  local total_cumulative_energy_imported = device:get_field(TOTAL_CUMULATIVE_ENERGY_IMPORTED_MAP) or {}
+  local total_energy = 0
+  for _, energyWh in pairs(total_cumulative_energy_imported) do
+    total_energy = total_energy + energyWh
+  end
+  return total_energy
+end
+
+local function schedule_energy_report_timer(device)
+  if not device:supports_capability(capabilities.powerConsumptionReport) then
+    return
+  end
+
+  local polling_schedule_timer = device:get_field(RECURRING_REPORT_TIMER)
+  if polling_schedule_timer ~= nil then
+    return
+  end
+
+  -- The powerConsumption report needs to be updated at least every 15 minutes in order to be included in SmartThings Energy
+  local pcr_interval = device:get_field(DEVICE_POWER_CONSUMPTION_REPORT_TIME_INTERVAL) or DEFAULT_REPORT_TIME_INTERVAL
+  pcr_interval = utils.clamp_value(pcr_interval, DEFAULT_REPORT_TIME_INTERVAL, MAX_REPORT_TIMEOUT)
+  local timer = device.thread:call_on_schedule(pcr_interval, function()
+    local last_time = device:get_field(LAST_REPORTED_TIME) or 0
+    local current_time = os.time()
+    local total_energy = get_total_cumulative_energy_imported(device)
+    device:set_field(LAST_REPORTED_TIME, current_time, { persist = true })
+
+    -- Calculate the energy consumed between the start and the end time
+    local previousTotalConsumptionWh = device:get_latest_state("main", capabilities.powerConsumptionReport
+      .ID,
+      capabilities.powerConsumptionReport.powerConsumption.NAME) or { energy = 0 }
+
+    local deltaEnergyWh = math.max(total_energy - previousTotalConsumptionWh.energy, 0.0)
+    local startTime = epoch_to_iso8601(last_time)
+    local endTime = epoch_to_iso8601(current_time - 1)
+
+    -- Report the energy consumed during the time interval. The unit of these values should be 'Wh'
+    device:emit_event(capabilities.powerConsumptionReport.powerConsumption({
+      start = startTime,
+      ["end"] = endTime,
+      deltaEnergy = deltaEnergyWh,
+      energy = total_energy
+    }))
+  end, "polling_report_schedule_timer")
+
+  device:set_field(RECURRING_REPORT_TIMER, timer)
+end
+
+local function delete_reporting_timer(device)
+  local reporting_poll_timer = device:get_field(RECURRING_REPORT_TIMER)
+  if reporting_poll_timer ~= nil then
+    device.thread:cancel_timer(reporting_poll_timer)
+    device:set_field(RECURRING_REPORT_TIMER, nil)
+  end
+end
+
+local function remove_timers(device)
+  delete_reporting_timer(device)
+  local poll_timer = device:get_field(RECURRING_POLL_TIMER)
+  if poll_timer ~= nil then
+    device.thread:cancel_timer(poll_timer)
+    device:set_field(RECURRING_POLL_TIMER, nil)
+  end
+end
```

Comment:
This seems like unnecessary breakdown of functionality to me. I would suggest:
```suggestion
local function remove_timers(device)
  -- remove report timer
  local reporting_poll_timer = device:get_field(RECURRING_REPORT_TIMER)
  if reporting_poll_timer ~= nil then
    device.thread:cancel_timer(reporting_poll_timer)
    device:set_field(RECURRING_REPORT_TIMER, nil)
  end
  -- remove poll timer
  local poll_timer = device:get_field(RECURRING_POLL_TIMER)
  if poll_timer ~= nil then
    device.thread:cancel_timer(poll_timer)
    device:set_field(RECURRING_POLL_TIMER, nil)
  end
end
```